sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala

```diff
@@ -114,6 +114,15 @@ private[sql] object TypeCollection {
     BooleanType,
     ByteType, ShortType, IntegerType, LongType)
 
+  /**
+   * Types that include numeric types and interval type. They are only used in unary_minus,
+   * unary_positive, add and subtract operations.
+   */
+  val NumericAndInterval = TypeCollection(
+    ByteType, ShortType, IntegerType, LongType,
+    FloatType, DoubleType, DecimalType,
+    IntervalType)
```

Comment:
Actually we should use `NumericType` here.
Think about `2 + "2"`, before this PR, we will first cast `"2"` to `2.0`(double is the default for numeric type), and then the result should be double 4.0. After this PR, we will first cast `"2"` to `2`(byte is the default for your type collection), and thus make the result of type int.